### PR TITLE
Create carbon_salbutamol

### PR DESCRIPTION
**Pull request by richiecroker**
first attempt at carbon inhaler measure

### CODE REVIEW

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  "denominator_short": "Salbutamol inhalers",
+  "why_it_matters": [
+    "The NHS has <a href='https://www.longtermplan.nhs.uk/wp-content/uploads/2019/08/nhs-long-term-plan-version-1.2.pdf'> ",
+    "committed to reducing its carbon footprint by 51% by 2025</a> to meet the target in the Climate Change Act, including a shift to lower carbon inhalers",
+    "to deliver a reduction of 4%.</p>",
+    "<p>Salbutamol metered dose inhalers (MDIs) are the single biggest source of carbon emissions from NHS medicines prescribing.",
+    "The 2022-2023 PCN Investment and Impact Fund (IIF) therefore ",
+    "<a href='https://www.england.nhs.uk/publication/network-contract-directed-enhanced-service-investment-and-impact-fund-2022-23-updated-guidance/'>",
+    "includes an indicator</a> focusing on mean carbon emissions per salbutamol inhaler prescribed (kg CO<sub>2</sub>e).",
+    "The aim is to encourage a change in prescribing of salbutamol devices to those associated with lower carbon emissions, resulting in a significant reduction in NHS carbon emissions.</p>",
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:19]: Reading this  - I'm wondering does the use of `devices` here imply a switch between different types of devices?

Comment (richiecroker) [reply openprescribing/measure_definitions/carbon_salbutamol.json:19]: we've pondered over this, and I think thought was reasonable wording.  We could have:
>The aim is to encourage prescribing of salbutamol products associated with lower carbon emissions, resulting in a significant reduction in NHS carbon emissions.

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    "<p>Data on carbon emissions per salbutamol inhaler can be viewed in the ",
+    "<a href='https://docs.google.com/spreadsheets/d/1eyk2iqNM_4sobDzevZWWcRR00O2tmJO5HwdFZLjhfa4'>",
+    "table here</a>.</p>"
+  ],
+  "tags": [
+    "core",
+    "respiratory"
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:26]: for a later date but we should consider reinstalling  https://www.bennett.ox.ac.uk/blog/2020/02/nhs-england-investment-and-impact-fund-a-new-dashboard-on-openprescribing/

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  ],
+  "url": null,
+  "is_percentage": false,
+  "is_cost_based": false,
+  "low_is_good": true,
+  "numerator_type": "custom",
+  "numerator_columns": [
+    "SUM(co2) AS numerator"
+  ],
+  "numerator_from": "{measures}.vw__carbon_salbutamol",
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:36]: Is there a way we could expose the code that makes up the view? It is essential for the measure.

Comment (richiecroker) [reply openprescribing/measure_definitions/carbon_salbutamol.json:36]: not without rewriting the measure

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  "numerator_where": [
+    "1 = 1"
+  ],
+  "numerator_bnf_codes_query": "SELECT DISTINCT bnf_code FROM {hscic}.presentation WHERE bnf_code IN ('0301011R0AAAPAP','0301011R0AABUBU','0301011R0AAAQAQ','0301011R0BEAHAQ','0301011R0BEAIAP','0301011R0BIAFAP','0301011R0BIAHAP','0301011R0BIAGBU','0301011R0BMAAAP','0301011R0BMABBU','0301011R0BWAABZ','0301011R0BWABAQ','0301011R0BXAACB','0301011R0BXAACB')",
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:40]: we should set a reminder to review this semi-regularly for new products

**Review by brianmackenna — APPROVED** [review]
This looks good to me but good to have a thorough tech review to ensure view is working like we'd expect. My two main bits of feedback
- can we expose view to make it open
- how will we trigger routine review to ensure no new products/bnf codes?

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  ],
+  "numerator_from": "{hscic}.normalised_prescribing",
+  "numerator_where": [
+    "bnf_code IN ('0301011R0AAAPAP','0301011R0AABUBU','0301011R0AAAQAQ','0301011R0BEAHAQ','0301011R0BEAIAP','0301011R0BIAFAP','0301011R0BIAHAP','0301011R0BIAGBU','0301011R0BMAAAP','0301011R0BMABBU','0301011R0BWAABZ','0301011R0BWABAQ','0301011R0BXAACB','0301011R0BXAACB')"
+  ],
+  "denominator_type": "custom",
+  "denominator_columns": [
+    "SUM(quantity) AS denominator"
+  ],
+  "denominator_from": "{hscic}.normalised_prescribing",
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:59]: why are we specifying from here? should it not be off defaults as per wiki?

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  "denominator_where": [
+    "bnf_code IN ('0301011R0AAAPAP','0301011R0AABUBU','0301011R0AAAQAQ','0301011R0BEAHAQ','0301011R0BEAIAP','0301011R0BIAFAP','0301011R0BIAHAP','0301011R0BIAGBU','0301011R0BMAAAP','0301011R0BMABBU','0301011R0BWAABZ','0301011R0BWABAQ','0301011R0BXAACB','0301011R0BXAACB')"
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:61]: Not sure if it matters but should we not be using `"numerator_bnf_codes_filter"` ? 

This is wiki specifies.

File: openprescribing/measure_definitions/carbon_salbutamol.json (hunk 5 above)
Comment (brianmackenna) [reply openprescribing/measure_definitions/carbon_salbutamol.json:59]: oh wait - we have to specify if custom denom, which this is

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    "WHEN bnf_code = '0301011R0BEAIAP' THEN quantity * 28.26 --Ventolin 100micrograms/dose Evohaler \n", 
+    "WHEN bnf_code = '0301011R0BIAFAP' THEN quantity * 11.95 --Salamol 100micrograms/dose inhaler CFC free (Teva) \n",
+    "WHEN bnf_code = '0301011R0BIAHAP' THEN quantity * 11.95 --Salamol 100micrograms/dose inhaler CFC free (Arrow) \n",
+    "WHEN bnf_code = '0301011R0BIAGBU' THEN quantity * 12.08 --Salamol 100micrograms/dose Easi-Breathe inhaler \n",
+    "WHEN bnf_code = '0301011R0BMAAAP' THEN quantity * 9.72 --Airomir 100micrograms/dose inhaler \n",
+    "WHEN bnf_code = '0301011R0BMABBU' THEN quantity * 9.72 --Airomir Autohaler \n",
+    "WHEN bnf_code = '0301011R0BWAABZ' THEN quantity * 0.62 --Easyhaler Salbutamol sulfate 100micrograms/dose dry pdr inh \n",
+    "WHEN bnf_code = '0301011R0BWABAQ' THEN quantity * 0.62 --Easyhaler Salbutamol sulfate 200micrograms/dose dry pdr inh \n",
+    "WHEN bnf_code = '0301011R0BXAACB' THEN quantity * 3.75 --Salbulin Novolizer 100micrograms/dose inhalation powder \n",
+    "WHEN bnf_code = '0301011R0BXAACB' THEN quantity * 3.75 --Salbulin Novolizer 100micrograms/dose inhalation powder refill \n",
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:48]: BNF code on line 47 and 48 repeat themselves.

I think this should be `0301011R0BXABCC`

File: openprescribing/measure_definitions/carbon_salbutamol.json (hunk 6 above)
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:61]: Last code in line 61 similar to above


I think this should be `0301011R0BXABCC`

File: openprescribing/measure_definitions/carbon_salbutamol.json [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    "ELSE 0 end) AS numerator"
+  ],
+  "numerator_from": "{hscic}.normalised_prescribing",
+  "numerator_where": [
+    "bnf_code IN ('0301011R0AAAPAP','0301011R0AABUBU','0301011R0AAAQAQ','0301011R0BEAHAQ','0301011R0BEAIAP','0301011R0BIAFAP','0301011R0BIAHAP','0301011R0BIAGBU','0301011R0BMAAAP','0301011R0BMABBU','0301011R0BWAABZ','0301011R0BWABAQ','0301011R0BXAACB','0301011R0BXAACB')"
```
Comment (brianmackenna) [inline openprescribing/measure_definitions/carbon_salbutamol.json:53]: Last code on line 53



I think this should be `0301011R0BXABCC`

**Review by brianmackenna — CHANGES_REQUESTED** [review]
Thanks for rewriting and exposing SQL query for open inspection - some nits above.

I'd really value a review from @inglesp / @evansd as well. I'm not as familiar reviewing `custom` measures and unsure if it could be optimised.

One final think I think we should do is write a test/notebook that is routinely checked to see if there are new products that might be relevant (this shouldn't stop us deploying the measure though).

**Review by inglesp — APPROVED** [review]
You don't need a `denominator_type` to be `"custom"` -- you could have:

```
    "denoninator_type": "bnf_quantity",
    "denominator_bnf_codes": [your list of BNF codes here]",
```

Apart from that, looks sensible.

**Comment by richiecroker** [comment]
thanks @inglesp , in this case the drugs are the same in both `numerator` and `denominator` so I mirrored them, so in case of change the code can be cut and pasted.  Is one more optimised than the other?

**Review by brianmackenna — APPROVED** [review]
Thanks!

File: openprescribing/measure_definitions/carbon_salbutamol.json (hunk 3 above)
Comment (brianmackenna) [reply openprescribing/measure_definitions/carbon_salbutamol.json]: done! good work

**Comment by brianmackenna** [comment]
This is great - so is the fix 
a) automated update
or
b) manual update and run this code every so often to ensure its up to date

**Comment by brianmackenna** [comment]
separately @chrisjwood16 would you be able to add a column to @richiecroker table above that covers possible/suggested match for carbon impact?

I think some will have no matches but some we can probably reasonable match as small changes in listed name

**Comment by richiecroker** [comment]
SQL and table in comment above fixed to remove duplication

**Comment by richiecroker** [comment]
Following update of code to include derived codes from brands, new check SQL
```SQL
SELECT rx.bnf_name,
            rx.bnf_code,
            SUM(items) AS items
FROM  hscic.normalised_prescribing AS rx
INNER JOIN measures.dmd_objs_with_form_route AS dmd
      ON rx.bnf_code = dmd.bnf_code
WHERE  dmd.form_route IN ('pressurizedinhalation.inhalation',
                          'powderinhalation.inhalation',
                          'inhalationsolution.inhalation')
AND dmd.bnf_code LIKE "0301011R0%"
AND dmd.bnf_code NOT IN ('0301011R0AAAAAA',
                         '0301011R0AABZBZ',
                         '0301011R0AACBCB',
                         '0301011R0AACCCC',
                         '0301011R0AAAPAP',
                         '0301011R0AABUBU',
                         '0301011R0AAAQAQ',
                         '0301011R0BEAHAQ',
                         '0301011R0BEAIAP',
                         '0301011R0BIAFAP',
                         '0301011R0BIAHAP',
                         '0301011R0BIAGBU',
                         '0301011R0BMAAAP',
                         '0301011R0BMABBU',
                         '0301011R0BWAABZ',
                         '0301011R0BWABAQ',
                         '0301011R0BXAACB',
                         '0301011R0BXABCC')
AND month BETWEEN '2017-01-01' AND '2022-02-01'
AND dmd.obj_type IN ('vmp', 'amp')
GROUP BY rx.bnf_name,
         rx.bnf_code
ORDER BY items DESC
```

gives:

| bnf_name                                                     | bnf_code        | items  |
|--------------------------------------------------------------|-----------------|--------|
| AirSalb 100micrograms/dose inhaler CFC   free                | 0301011R0BZAAAP | 1007   |
| Salbutamol 95micrograms/dose dry powder   inhaler            | 0301011R0AAAFAF | 527   |
| Asmasal 95micrograms/dose Clickhaler                         | 0301011R0BJABAF | 178    |
| Asmavent 100micrograms/dose inhaler CFC   free               | 0301011R0BYAAAP | 4      |
| Salbutamol 100micrograms/dose breath   actuated inhaler      | 0301011R0AAADAD | 4      |
| Pulvinal Salbutamol 200micrograms/dose   dry powder inhaler  | 0301011R0BTAAAQ | 3      |
| Ventodisks 400microgram                                      | 0301011R0BLADAL | 1      |
| Ventodisks 200microgram                                      | 0301011R0BLABAK | 1      |

These are all now discontinued, so no need to map them for carbon footprint.

**Comment by richiecroker** [comment]
I've also amended notation here: https://docs.google.com/spreadsheets/d/1eyk2iqNM_4sobDzevZWWcRR00O2tmJO5HwdFZLjhfa4/edit#gid=0

**Pull request merged**